dlt/extract/incremental.py

```diff
@@ -0,0 +1,133 @@
+        self._state = state
+
+    @classmethod
+    def from_config(cls, cfg: IncrementalConfigSpec, resource_name: str) -> "Incremental":
+        # TODO: last_value_func from name
+        return cls(cfg.cursor_column, initial_value=cfg.initial_value, resource_name=resource_name)
+
+    @property
+    def state(self) -> IncrementalColumnState:
+        return self._state().setdefault('incremental', {}).setdefault(
+            self.resource_name, {}
+        ).setdefault(self.cursor_column, {'last_value': self.initial_value, 'unique_keys': []})
+
+    @property
+    def last_value(self) -> Optional[TCursorValue]:
+        return self.state['last_value']
+
+    @property
+    def unique_keys(self) -> List[TUniqueValue]:
+        return self.state['unique_keys']
+
+    def __call__(self, row: TDataItem) -> bool:
+        # TODO: A bit inefficient to run this for each row.
+        # Can we use a transform which accepts a list of rows when resource yields in batches?
+        if row is None:
+            # TODO: should this fail or just return?
+            pass
+        state = self.state
+        last_value = state['last_value']
+        # TODO: Should it be an error if row missing the cursor?
+        # TODO: Extract value by json-path
+        row_value = row[self.cursor_column]
+        if self.primary_key:
+            unique_value = row[self.primary_key]
+        else:
+            # TODO: Need to think about edge cases.
+            # e.g. if schema has changed since last run, row may have new columns
+            # do we then still want to load a duplicate?
+            unique_value = sha1(json.dumpb(row, sort_keys=True)).hexdigest()
+        if last_value == row_value:
+            if unique_value in state['unique_keys']:
+                return False
+            state['unique_keys'].append(unique_value)
+            return True
+        new_value = self.last_value_func(last_value, row_value) if last_value is not None else row_value
+        if new_value != last_value:
+            state.update({'last_value': new_value, 'unique_keys': [unique_value]})
+        return True
+
+    def copy(self):
+        return Incremental(self.cursor_column)
+
+
+incremental = Incremental
+
+
+
+def with_incremental(func: TFun, resource_name: str, resource_sections: Tuple[str, ...]) -> TFun:
+    @wraps(func)
+    def _wrap(*args, **kwargs):
+        sig = inspect.signature(func)
+        new_kwargs = {}
+
+        for p in sig.parameters.values():
+            if p.name in kwargs:
+                value = kwargs[p.name]
+                value.resource_name = resource_name
+                new_kwargs[p.name] = value
+            elif issubclass(p.annotation, Incremental) or isinstance(p.default, Incremental):
+                cfg = resolve_configuration(IncrementalConfigSpec(), sections=resource_sections, explicit_value=p.default)
+                new_kwargs[p.name] = Incremental.from_config(cfg, resource_name)
```

Comment:
Yeah just wanted to see if this would work first, and it does. Config resolution in dlt is powerful :D  
Next is to put all these ideas together.

Do you mean to put the `with_incremental` wrapping in `DltResource`?  
I was doing it in the decorator because the config sections were already defined there.  

I'm thinking `with_incremental` could resolve the `Incremental` instance outside the wrapper and return both the wrapped func and the instance:

```
def with_incremental(func: TFun, ...) -> Tuple[TFun, Incremental]:
    ...
```

And the @resource decorator simply then passes a completely configured `Incremental` object to `DltResource`:

```
conf_f, incremental = with_incremental(conf_f, ...)

...
# make_resource

DltResource.from_data(..., incremental=incremental)
```

`DltResource` then simply receives a fully configured incremental object and only has to add the pipe step.